File: Semester 5/Mobile Applications/non_native_server_support/non_native/lib/Model/Character.dart
class Character {
  String characterId;
  String name;
  String story;
  String description;
  String age;
  String background;
  String skills;
  String strengths;
  String weaknesses;
  String notableFeatures;
  bool isFavourite;

  Character({
    this.name = "",
    this.story = "",
    this.description = "",
    this.age = "",
    this.background = "",
    this.skills = "",
    this.strengths = "",
    this.weaknesses = "",
    this.notableFeatures = "",
    this.isFavourite = false,
    this.characterId = "-1",
  });

  Map<String, dynamic> toMapWithID() {
    return {
      'characterId': characterId,
      'name': name,
      'story': story,
      'description': description,
      'age': age,
      'background': background,
      'skills': skills,
      'strengths': strengths,
      'weaknesses': weaknesses,
      'notableFeatures': notableFeatures,
      'isFavourite': isFavourite ? 1 : 0
    };
  }

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'story': story,
      'description': description,
      'age': age,
      'background': background,
      'skills': skills,
      'strengths': strengths,
      'weaknesses': weaknesses,
      'notableFeatures': notableFeatures,
      'isFavourite': isFavourite ? 1 : 0
    };
  }

  factory Character.fromMap(Map<String, dynamic> map) {
    var character = Character(
      name: map['name'] ?? '',
      story: map['story'] ?? '',
      description: map['description'] ?? '',
      age: map['age'] ?? '',
      background: map['background'] ?? '',
      skills: map['skills'] ?? '',
      strengths: map['strengths'] ?? '',
      weaknesses: map['weaknesses'] ?? '',
      notableFeatures: map['notableFeatures'] ?? '',
      isFavourite: map['isFavourite'] == 1,
    );
    character.characterId = map['characterId'];
    return character;
  }

  Map<String, dynamic> toJson() {
    return {
      'characterId': characterId,
      'name': name,
      'story': story,
      'description': description,
      'age': age,
      'background': background,
      'skills': skills,
      'strengths': strengths,
      'weaknesses': weaknesses,
      'notableFeatures': notableFeatures,
      'isFavourite': isFavourite,
    };
  }

  factory Character.fromJson(Map<String, dynamic> map) {
    return Character(
      characterId:  map['characterId'] is String
          ? map['characterId']
          : map['characterId'].toString(),
      name: map['name'],
      story: map['story'],
      description: map['description'],
      age: map['age'],
      background: map['background'],
      skills: map['skills'],
      strengths: map['strengths'],
      weaknesses: map['weaknesses'],
      notableFeatures: map['notableFeatures'],
      isFavourite: false,
    );
  }

  @override
  String toString() {
    return 'Character(name: $name, story: $story, description: $description, age: $age, characterId: $characterId)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! Character) return false;
    return characterId == other.characterId &&
        name == other.name &&
        story == other.story &&
        description == other.description &&
        age == other.age &&
        background == other.background &&
        skills == other.skills &&
        strengths == other.strengths &&
        weaknesses == other.weaknesses &&
        notableFeatures == other.notableFeatures &&
        isFavourite == other.isFavourite;
  }
}

File: Semester 5/Mobile Applications/non_native_server_support/server/src/main/java/com/server/server/Model/Character.java
package com.server.server.Model;

import com.google.gson.Gson;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@SuppressWarnings("ALL")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
public class Character {
    @Id
    public String characterId;
    private String name;
    private String story;
    private String description;
    private String age;
    private String background;
    private String skills;
    private String strengths;
    private String weaknesses;
    private String notableFeatures;
    private boolean isFavourite;

    public String toJson() {
        return new Gson().toJson(this);
    }

    public String getCharacterId() {
        return characterId;
    }

    public void setCharacterId(String characterId) {
        this.characterId = characterId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getStory() {
        return story;
    }

    public void setStory(String story) {
        this.story = story;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getAge() {
        return age;
    }

    public void setAge(String age) {
        this.age = age;
    }

    public String getBackground() {
        return background;
    }

    public void setBackground(String background) {
        this.background = background;
    }

    public String getSkills() {
        return skills;
    }

    public void setSkills(String skills) {
        this.skills = skills;
    }

    public String getStrengths() {
        return strengths;
    }

    public void setStrengths(String strengths) {
        this.strengths = strengths;
    }

    public String getWeaknesses() {
        return weaknesses;
    }

    public void setWeaknesses(String weaknesses) {
        this.weaknesses = weaknesses;
    }

    public String getNotableFeatures() {
        return notableFeatures;
    }

    public void setNotableFeatures(String notableFeatures) {
        this.notableFeatures = notableFeatures;
    }

    public boolean isFavourite() {
        return isFavourite;
    }

    public void setFavourite(boolean isFavourite) {
        this.isFavourite = isFavourite;
    }

}
